### PR TITLE
Upgrade rubocop rspec gem

### PR DESCRIPTION
**Pull request by jonas054**
I think I was able to fix the newly detected offenses most of the time, but had to resort to `rubocop:disable` in a few cases. There is some duplicated code that's been removed, so that has to be reviewed carefully.

### CODE REVIEW

File: spec/rubocop/cop/team_spec.rb [hunk 1]
```diff
@@ -368,13 +368,15 @@ def foo
...
       before do
+        # rubocop:disable RSpec/LeakyConstantDeclaration
         module Test
           class CopWithExternalDeps < ::RuboCop::Cop::Cop
             def external_dependency_checksum
               'something other than nil'
             end
           end
         end
+        # rubocop:enable RSpec/LeakyConstantDeclaration
```
Comment (pirj) [inline spec/rubocop/cop/team_spec.rb:379]: A quick and dirty to overcome this inconvenience with defining cops in specs with `Class.new`:
```
diff --git a/lib/rubocop/cop/registry.rb b/lib/rubocop/cop/registry.rb
index 150dc29e0..e311b8fda 100644
--- a/lib/rubocop/cop/registry.rb
+++ b/lib/rubocop/cop/registry.rb
@@ -23,8 +23,7 @@ module RuboCop
     # Registry that tracks all cops by their badge and department.
     class Registry
       def initialize(cops = [], options = {})
-        @registry = {}
-        @departments = {}
+        @lazy = []
         @cops_by_cop_name = Hash.new { |hash, key| hash[key] = [] }
 
         cops.each { |cop| enlist(cop) }
@@ -32,10 +31,7 @@ module RuboCop
       end
 
       def enlist(cop)
-        @registry[cop.badge] = cop
-        @departments[cop.department] ||= []
-        @departments[cop.department] << cop
-        @cops_by_cop_name[cop.cop_name] << cop
+        @lazy << cop
       end
 
       # @return [Array<Symbol>] list of departments for current cops.
@@ -129,12 +125,28 @@ module RuboCop
         @cops_by_cop_name
       end
 
+      def registry
+        @registry ||= init_registry
+      end
+
+      def init_registry
+        @registry = {}
+        @departments = {}
+        @lazy.each do |cop|
+          @registry[cop.badge] = cop
+          @departments[cop.department] ||= []
+          @departments[cop.department] << cop
+          @cops_by_cop_name[cop.cop_name] << cop
+        end
+        @registry
+      end
+
       def cops
-        @registry.values
+        registry.values
       end
 
       def length
-        @registry.size
+        registry.size
       end
 
       def enabled(config, only, only_safe = false)
@@ -172,7 +184,7 @@ module RuboCop
       end
 
       def sort!
-        @registry = Hash[@registry.sort_by { |badge, _| badge.cop_name }]
+        @registry = Hash[registry.sort_by { |badge, _| badge.cop_name }]
 
         self
       end
@@ -214,7 +226,7 @@ module RuboCop
       end
 
       def registered?(badge)
-        @registry.key?(badge)
+        registry.key?(badge)
       end
     end
   end
```

Comment (pirj) [reply spec/rubocop/cop/team_spec.rb:379]: I understand that the only reason to change how it is done now is to make tests simpler/more conformant.

File: spec/rubocop/target_finder_spec.rb [hunk 2]
```diff
@@ -78,6 +20,64 @@
     create_file('dir1/executable',  '#!/usr/bin/env ruby')
     create_empty_file('dir2/ruby3.rb')
     create_empty_file('.hidden/ruby4.rb')
+
+    stub_const('RUBY_EXTENSIONS', %w[.rb
```
Comment (pirj) [inline spec/rubocop/target_finder_spec.rb:24]: Same as in one of the above cases - a variable, `let` or a method would be preferrable. `stub_const` is unnecessary.

File: spec/rubocop/config_spec.rb [hunk 3]
```diff
@@ -150,33 +150,6 @@
...
-            EnforcedStyleInsidePipes: itisinvalid
-        YAML
-      end
-
-      it 'raises validation error' do
-        expect { configuration.validate }
-          .to raise_error(RuboCop::ValidationError, /itisinvalid/)
-      end
-    end
-
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/config_spec.rb:179]: Right above these two removed contexts there are two other ones with the same descriptions. They use another cop in configuration, and I see now that there is a difference between them. In the first examples, the parameter name is `EnforcedStyle`. In the removed ones it's still an an enforced style, but the parameter name is `EnforcedStyleInsidePipes`. So I think I'll take back these examples and just update the descriptions instead.

File: spec/rubocop/cop/layout/space_inside_block_braces_spec.rb [hunk 4]
```diff
@@ -1,10 +1,12 @@
 # frozen_string_literal: true
 
 RSpec.describe RuboCop::Cop::Layout::SpaceInsideBlockBraces, :config do
-  SUPPORTED_STYLES = %w[space no_space].freeze
-
   subject(:cop) { described_class.new(config) }
 
+  before do
+    stub_const('SUPPORTED_STYLES', %w[space no_space].freeze)
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/cop/layout/space_inside_block_braces_spec.rb:7]: Good idea. Will do.

File: spec/rubocop/cop/registry_spec.rb [hunk 5]
```diff
@@ -22,6 +23,7 @@ class Foo < Cop
         end
       end
     end
+    # rubocop:enable RSpec/LeakyConstantDeclaration
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/cop/registry_spec.rb:26]: Yes, that's probably it. I tried this to begin with, but couldn't get it to work.

File: spec/rubocop/cop/style/each_for_simple_loop_spec.rb [hunk 6]
```diff
@@ -3,8 +3,10 @@
 RSpec.describe RuboCop::Cop::Style::EachForSimpleLoop do
   subject(:cop) { described_class.new }
 
-  OFFENSE_MSG = 'Use `Integer#times` for a simple loop ' \
-                'which iterates a fixed number of times.'
+  # before do
+  #   stub_const('OFFENSE_MSG', 'Use `Integer#times` for a simple loop ' \
+  #                             'which iterates a fixed number of times.')
+  # end
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/cop/style/each_for_simple_loop_spec.rb:9]: Oops! Good catch.

File: spec/rubocop/cop/style/multiline_method_signature_spec.rb [hunk 7]
```diff
@@ -49,7 +49,7 @@ def foo
     end
   end
 
-  context 'when arguments span multiple lines' do
+  context 'when arguments for a singleton method span multiple lines' do
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/cop/style/multiline_method_signature_spec.rb:52]: True. I'll re-arrange.

File: spec/rubocop/cop/style/trailing_comma_in_arguments_spec.rb [hunk 8]
```diff
@@ -441,15 +441,15 @@
       end
 
       it 'does not break when a safe method call is chained on the ' \
-         'offending one' do
+         'offending simple one' do
```
[the remark this answers is not in the document]
Comment (jonas054) [reply spec/rubocop/cop/style/trailing_comma_in_arguments_spec.rb:444]: Yes. This group of examples were added when fixing #6755. I just tweaked the wording of a duplicated description.

File: spec/rubocop/cop/team_spec.rb (hunk 1 above)
Comment (jonas054) [reply spec/rubocop/cop/team_spec.rb:379]: Cool! Would you make that into a PR?

File: spec/rubocop/target_finder_spec.rb (hunk 2 above)
Comment (jonas054) [reply spec/rubocop/target_finder_spec.rb:24]: :+1: Will fix.

**Comment by jonas054** [comment]
@pirj Thank for the review! I've pushed some updates.

File: spec/rubocop/cop/layout/space_inside_block_braces_spec.rb [hunk 9]
```diff
@@ -3,14 +3,11 @@
 RSpec.describe RuboCop::Cop::Layout::SpaceInsideBlockBraces, :config do
   subject(:cop) { described_class.new(config) }
 
-  before do
-    stub_const('SUPPORTED_STYLES', %w[space no_space].freeze)
-  end
-
+  let(:supported_styles) { %w[space no_space].freeze }
```
Comment (pirj) [inline spec/rubocop/cop/layout/space_inside_block_braces_spec.rb:6]: No need to freeze, it's a throwaway `let`.

File: spec/rubocop/target_finder_spec.rb [hunk 10]
```diff
@@ -11,6 +11,67 @@
...
+       .rake
+       .rbuild
+       .rbw
+       .rbx
+       .ru
+       .ruby
+       .spec
+       .thor
+       .watchr].freeze
+  end
```
Comment (pirj) [inline spec/rubocop/target_finder_spec.rb:39]: Same here.
I'd go with a regular variable. Just to blow people's minds and remind that RSpec is not a separate language, but just Ruby.

File: spec/rubocop/cop/team_spec.rb (hunk 1 above)
Comment (pirj) [reply spec/rubocop/cop/team_spec.rb:379]: Here you go https://github.com/rubocop-hq/rubocop/pull/8023

File: spec/rubocop/target_finder_spec.rb (hunk 10 above)
Comment (jonas054) [reply spec/rubocop/target_finder_spec.rb:39]: Well said. :slightly_smiling_face:

**Pull request merged**